// format-package/build/cli/parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = void 0;
const yargs = require("yargs");
const parser = yargs
    .command(['format [files..]', '*'], 'Format files', (commandYargs) => commandYargs.positional('files', {
    default: ['**/package.json'],
    describe: 'Files to be formatted (accepts globs)',
    type: 'string',
    alias: 'globs',
}))
    .env('FORMAT_PACKAGE')
    .options({
    config: {
        alias: 'c',
        demandOption: false,
        describe: `Location of a config file`,
        type: 'string',
    },
    write: {
        alias: 'w',
        demandOption: false,
        default: false,
        describe: 'Flag to write the output of ordering to the package.json file',
        type: 'boolean',
    },
    verbose: {
        alias: 'v',
        demandOption: false,
        default: false,
        describe: 'Flag to make the script print the formatted result',
        type: 'boolean',
    },
    ignore: {
        alias: 'i',
        demandOptions: false,
        default: ['**/node_modules/**'],
        describe: 'Patterns to ignore',
        type: 'array',
    },
})
    .alias('h', 'help')
    .help()
    .strict();
const parse = (argv) => {
    // Safely cast
    return parser.parse(argv);
};
exports.default = parse;
